File: apps/frontend/src/adapters/shared/photos.ts
import { invoke } from "./platform";

export interface FilmRoll {
  id: string;
  name: string;
  filmTypeKey: string;
  artworkKey: string;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
}

export interface FilmRollSummary extends FilmRoll {
  photoCount: number;
}

export interface NewFilmRoll {
  id?: string;
  name: string;
  filmTypeKey?: string;
  artworkKey?: string;
  sortOrder?: number;
}

export interface UpdateFilmRoll {
  name?: string;
  filmTypeKey?: string;
  artworkKey?: string;
  sortOrder?: number;
}

export type DeleteFilmRollMode = "MovePhotosToTray" | "DeletePhotos";

export interface Photo {
  id: string;
  filePath: string;
  originalFileName?: string | null;
  thumbnailPath?: string | null;
  filmRollId?: string | null;
  sortOrder: number;
  importedAt: string;
  createdAt: string;
  updatedAt: string;
}

export interface NewPhoto {
  id?: string;
  filePath: string;
  originalFileName?: string | null;
  thumbnailPath?: string | null;
  filmRollId?: string | null;
  sortOrder?: number;
}

export type TrayItem = { type: "FilmRoll"; item: FilmRollSummary } | { type: "Photo"; item: Photo };

export const listTrayItems = async (): Promise<TrayItem[]> => {
  return invoke<TrayItem[]>("list_tray_items");
};

export const listFilmRolls = async (): Promise<FilmRollSummary[]> => {
  return invoke<FilmRollSummary[]>("list_film_rolls");
};

export const getFilmRoll = async (filmRollId: string): Promise<FilmRoll | null> => {
  return invoke<FilmRoll | null>("get_film_roll", { filmRollId });
};

export const listFilmRollPhotos = async (filmRollId: string): Promise<Photo[]> => {
  return invoke<Photo[]>("list_film_roll_photos", { filmRollId });
};

export const createPhoto = async (input: NewPhoto): Promise<Photo> => {
  return invoke<Photo>("create_photo", { input });
};

export const createFilmRoll = async (input: NewFilmRoll): Promise<FilmRoll> => {
  return invoke<FilmRoll>("create_film_roll", { input });
};

export const updateFilmRoll = async (
  filmRollId: string,
  patch: UpdateFilmRoll,
): Promise<FilmRoll> => {
  return invoke<FilmRoll>("update_film_roll", { filmRollId, patch });
};

export const deleteFilmRoll = async (
  filmRollId: string,
  mode: DeleteFilmRollMode = "MovePhotosToTray",
): Promise<void> => {
  await invoke<void>("delete_film_roll", { filmRollId, mode });
};

export const movePhotos = async (
  photoIds: string[],
  destinationFilmRollId: string | null,
): Promise<void> => {
  await invoke<void>("move_photos", { photoIds, destinationFilmRollId });
};

File: apps/tauri/src/commands/photos.rs
use std::sync::Arc;

use crate::context::ServiceContext;
use log::debug;
use panorama_core::photos::{
    DeleteFilmRollMode, FilmRoll, FilmRollSummary, NewFilmRoll, NewPhoto, Photo, TrayItem,
    UpdateFilmRoll,
};
use tauri::State;

#[tauri::command]
pub async fn list_tray_items(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<TrayItem>, String> {
    debug!("Listing tray items...");
    state
        .photo_service()
        .list_tray_items()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_film_rolls(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<FilmRollSummary>, String> {
    debug!("Listing film rolls...");
    state
        .photo_service()
        .list_film_rolls()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_film_roll(
    film_roll_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Option<FilmRoll>, String> {
    debug!("Fetching film roll...");
    state
        .photo_service()
        .get_film_roll(&film_roll_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_film_roll_photos(
    film_roll_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<Photo>, String> {
    debug!("Listing film roll photos...");
    state
        .photo_service()
        .list_film_roll_photos(&film_roll_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_photo(
    input: NewPhoto,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Photo, String> {
    debug!("Creating photo...");
    state
        .photo_service()
        .create_photo(input)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_film_roll(
    input: NewFilmRoll,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<FilmRoll, String> {
    debug!("Creating film roll...");
    state
        .photo_service()
        .create_film_roll(input)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_film_roll(
    film_roll_id: String,
    patch: UpdateFilmRoll,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<FilmRoll, String> {
    debug!("Updating film roll...");
    state
        .photo_service()
        .update_film_roll(&film_roll_id, patch)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_film_roll(
    film_roll_id: String,
    mode: DeleteFilmRollMode,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<usize, String> {
    debug!("Deleting film roll...");
    state
        .photo_service()
        .delete_film_roll(&film_roll_id, mode)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn move_photos(
    photo_ids: Vec<String>,
    destination_film_roll_id: Option<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<usize, String> {
    debug!("Moving photos...");
    state
        .photo_service()
        .move_photos(&photo_ids, destination_film_roll_id.as_deref())
        .await
        .map_err(|e| e.to_string())
}
